commit 30dbdbe75384bcb034172c781b24bc982434a967
Author: Tom Lane <tgl@sss.pgh.pa.us>
Date:   Mon Mar 13 12:40:28 2023 -0400

    Fix failure to detect some cases of improperly-nested aggregates.
    
    check_agg_arguments_walker() supposed that it needn't descend into
    the arguments of a lower-level aggregate function, but this is
    just wrong in the presence of multiple levels of sub-select.  The
    oversight would lead to executor failures on queries that should
    be rejected.  (Prior to v11, they actually were rejected, thanks
    to a "redundant" execution-time check.)
    
    Per bug #17835 from Anban Company.  Back-patch to all supported
    branches.
    
    Discussion: https://postgr.es/m/17835-4f29f3098b2d0ba4@postgresql.org

diff --git a/src/backend/parser/parse_agg.c b/src/backend/parser/parse_agg.c
index 4fbf80c2713..85cd47b7aec 100644
--- a/src/backend/parser/parse_agg.c
+++ b/src/backend/parser/parse_agg.c
@@ -734,12 +734,11 @@ check_agg_arguments_walker(Node *node,
 		{
 			if (context->min_agglevel < 0 ||
 				context->min_agglevel > agglevelsup)
 				context->min_agglevel = agglevelsup;
 		}
-		/* no need to examine args of the inner aggregate */
-		return false;
+		/* Continue and descend into subtree */
 	}
 	if (IsA(node, GroupingFunc))
 	{
 		int			agglevelsup = ((GroupingFunc *) node)->agglevelsup;
 
diff --git a/src/test/regress/expected/aggregates.out b/src/test/regress/expected/aggregates.out
index e074cb71bf6..f0517f95b60 100644
--- a/src/test/regress/expected/aggregates.out
+++ b/src/test/regress/expected/aggregates.out
@@ -1280,10 +1280,16 @@ LINE 1: select max(min(unique1)) from tenk1;
                    ^
 select (select max(min(unique1)) from int8_tbl) from tenk1;
 ERROR:  aggregate function calls cannot be nested
 LINE 1: select (select max(min(unique1)) from int8_tbl) from tenk1;
                            ^
+select avg((select avg(a1.col1 order by (select avg(a2.col2) from tenk1 a3))
+            from tenk1 a1(col1)))
+from tenk1 a2(col2);
+ERROR:  aggregate function calls cannot be nested
+LINE 1: select avg((select avg(a1.col1 order by (select avg(a2.col2)...
+                                                        ^
 --
 -- Test removal of redundant GROUP BY columns
 --
 create temp table t1 (a int, b int, c int, d int, primary key (a, b));
 create temp table t2 (x int, y int, z int, primary key (x, y));
diff --git a/src/test/regress/sql/aggregates.sql b/src/test/regress/sql/aggregates.sql
index 616ef38c259..1783d19bd5e 100644
--- a/src/test/regress/sql/aggregates.sql
+++ b/src/test/regress/sql/aggregates.sql
@@ -437,10 +437,13 @@ select distinct min(f1), max(f1) from minmaxtest;
 drop table minmaxtest cascade;
 
 -- check for correct detection of nested-aggregate errors
 select max(min(unique1)) from tenk1;
 select (select max(min(unique1)) from int8_tbl) from tenk1;
+select avg((select avg(a1.col1 order by (select avg(a2.col2) from tenk1 a3))
+            from tenk1 a1(col1)))
+from tenk1 a2(col2);
 
 --
 -- Test removal of redundant GROUP BY columns
 --
 
